Dark mode

slowbackspace opened the pull request:
close https://github.com/trezor/trezor-suite/issues/2738
close https://github.com/trezor/trezor-suite/issues/2226
close https://github.com/trezor/trezor-suite/issues/2894
close https://github.com/trezor/trezor-suite/issues/2888
apparently it also fixes https://github.com/trezor/trezor-suite/issues/2793
apparently it also fixes https://github.com/trezor/trezor-suite/issues/2841

for another PR:
 - desktop: avoid flash of default theme (on mac with dark mode enabled you will still see the flash of light theme, even tho we can tell the OS is in dark mode even before running javascript), maybe we could take it further and save selected theme to a file, so we don't even have to wait for idb to load our favorite theme. I'm leaving this for another PR as it will take more time to find the best solution.

## tldr; on what happened:
- Cleaned up colors definition in `@trezor/components` (converted everything to colors from our current design, removed old colors). Added new `THEME` object which holds color scheme for each theme (`THEME = {light: {...}, dark: {} }`)

- Whole component tree is now wrapped with our custom `ThemeProvider`, which uses `ThemeProvider` from styled-components under the hood. Its purpose is to pass colors for selected theme to SC's ThemeProvider which automatically takes care of injecting theme prop to all styled components. There is also additional logic for automatically selecting the theme based on OS theme (tested on mac, it works!). 
**Then in the component itself, instead of doing `color: ${colors.NEUE_TYPE_DARK_GREY}` you simply use passed theme prop `color: ${props => props.theme.NEUE_TYPE_DARK_GREY}`. Now the color will change automatically with selected theme. The `theme` prop is also correctly typed, on hover you can see what are the values given color resolves in.
Sometimes you need to use a correct color outside of styled component, eg. typically passing color as a prop to a component in JSX (`<Icon color={...})`), to help you with this, there is `useTheme` hook which will return colors object for currently active theme. `const theme = useTheme(); ... <Icon color={theme.NEUE_TYPE_RED} ... />`**

- Theme setting is stored inside a  a suite reducer (`suite.settings.theme`) and it is synced to persistent storage. Themes defined by us (right now "light" and "dark") will only use `theme.variant` field. However in debug settings there is an option to pick your own colors for dark mode. For this case the `theme.variant` will be set to 'custom', and all custom colors will be stored in `theme.colors`.

- reworked the way we inject global styles to be more idiomatic, now leveraging the injected `theme` prop

- There are bunch of little things, such as for images we are lowering their brightness and increasing contrast otherwise they look washed-out (you might say that they still do, but that's designer's problem now!), scrollbars also gets darker (even in firefox), updated coin icons, fixed images that had transparent layer


ohh, btw while I was working on this I run into this awesome write-up on dark mode from UI/UX perspective https://css-tricks.com/a-complete-guide-to-dark-mode-on-the-web/

mroz22 commented:
I would suggest merging this rather soon and put it behind a feature flag. Otherwise you will die from resolving conflicts..

slowbackspace commented:
@mroz22 It will be fine :) I am gonna resolve them all at once, and also I'll update newly added components, once everything is ready for review (probably later today) which I hope will be quick.

slowbackspace commented:
🎉 🎉 🎉  I am marking this as ready for review, I am still waiting for few things from our beloved product team, but these will boring changes anyway so feel free to take a look (PR looks big, but these are mostly 'search-n-replace' type of changes, so don't be scared). There is certainly room for improvements which I would like to make in another PR(s). Mainly we could get of some colors definitions and derive them from existing colors using something like `color` lib and also find a way to solve  "flash of default theme" in desktop build.

I am sure there will be some things missing/visually broken (most likely dark text on dark bg) so the sooner we can merge it into the develop the better is our chance to catch these issues before next release :)

keraf reviewed (changes requested):
Glanced over the functional changes of this PR and code looks mostly good to me. Just a few comments.

I found an issue with the toggling of the dark/light freezing the window for a few milliseconds on Firefox (haven't checked in Chrome). This is from the performance tab when toggling it on and off.
![Screenshot from 2020-11-12 15-24-25](https://user-images.githubusercontent.com/30026749/98953217-b01d4980-24fc-11eb-9c70-48cb2f73bc44.png).

Regarding the design, I'm not sure if we really need to have the dark-mode switch button in the nav bar. It feels to me like it's a setting that once enabled will rarely get fiddled with. Detection is automatic and I feel we should only have it in the settings. I would also maybe suggest changing it into a list instead of a toggle and change the wording around theming. It would be nice to have more themes in the future (community contributions, secret unlockable or even accessibility focused theme) and would show also that we're open to the idea of adding more.

I've also noticed the analytics is missing. It would be interesting to see if people use themes or not.

mroz22 commented:
> Regarding the design, I'm not sure if we really need to have the dark-mode switch button in the nav bar. It feels to me like it's a setting that once enabled will rarely get fiddled with

My words. I'd say this can go into settings. But who am I to judge designers choices?
And actually, for now, the switch on a reachable position may be helpful to quickly validate that there are no unexpected differences between normal and dark mode.

mroz22 commented on packages/components/src/components/Icon/index.tsx:
```diff
@@ -102,6 +102,8 @@ const Icon = React.forwardRef(
         ref?: React.Ref<HTMLDivElement>
     ) => {
+        const theme = useTheme();
+        const defaultColor = color ?? theme.TYPE_LIGHT_GREY;
```
probably good idea to adapt mentally to using `??` everywhere instead of `||`

mroz22 commented on packages/components/src/components/buttons/Pin/index.tsx:
```diff
@@ -1,30 +1,30 @@
-    height: 80px;
-    border: 1px solid ${colors.BLACK0};
-    background: ${colors.WHITE};
+    max-width: 100px;
+    max-height: 100px;
     transition: all 0.3s;
     position: relative;
     cursor: pointer;
+    margin: 4px;
+
+    padding-top: 30%;
+    width: 100%;
+
+    border-radius: 4px;
+    border: 1px solid ${props => props.theme.STROKE_GREY};
+    background: ${props => props.theme.BG_LIGHT_GREY};
```
just ideas, haven't checked yet, but:
1. doesn't it break some nested modal pin window?
1. does it keep buttons squared?

slowbackspace replied:
ohh, it should be the same. I noticed there were two pin button components. One in suite and one in trezor/components. We were using one from suite, so I copied its guts to the trezor/component's pin button and switched the import (and ofc removed no longed used pin button from suite).

mroz22 commented on packages/suite/src/components/suite/NavigationBar/components/NavigationActions/index.tsx:
```diff
@@ -132,6 +136,15 @@ const NavigationActions = (props: Props) => {
+                        setTheme(theme.variant === 'light' ? 'dark' : 'light');
+                    }}
+                    isActive={theme.variant === 'dark'}
+                    label={<Translation id="TR_DISCREET" />}
```
This is one of the todos you mentioned for a future PR, right?

slowbackspace replied:
this will get fixed (removed) in this PR :) We are going to need only switch in the settings.

mroz22 commented on packages/suite/src/components/suite/PasswordStrengthIndicator/index.tsx:
```diff
@@ -10,7 +10,7 @@ interface WrapperProps {
     display: flex;
     margin: 8px 0;
-    flex: 1;
+    /* flex: 1; */
```
everything works nice even after commenting it out?

slowbackspace replied:
huh, indeed that was probably not necessary. I noticed the strength indicator wasn't showing at all (latest develop), so I tried to figure out what happened to it. This is probably some left over from this.

mroz22 commented on packages/suite/src/components/suite/TransactionsGraph/components/RangeSelector/index.tsx:
```diff
@@ -15,7 +15,7 @@ const RangeItem = styled.div<{ selected: boolean }>`
     text-align: center;
     font-weight: ${props => (props.selected ? 600 : 500)};
-    color: ${props => (props.selected ? colors.NEUE_TYPE_DARK_GREY : colors.NEUE_TYPE_LIGHT_GREY)};
+    color: ${props => (props.selected ? props.theme.TYPE_DARK_GREY : props.theme.TYPE_LIGHT_GREY)};
```
does props.theme exist here?

mroz22 commented on packages/suite/src/support/suite/ThemeProvider.tsx:
```diff
@@ -0,0 +1,66 @@
+                ...theme.colors, // custom saved colors
+            };
+        default:
+            return THEME.light;
```
I am thinking about whether it won't be better to do migration and fill store with theme default. But this is probably good enough as long as somebody does not remove this default.

slowbackspace replied:
yeah, if we are gonna do some changes we are supposed to do proper migration. this is just for the events when we forget about it :)

slowbackspace replied:
ohh wait. you werent talking about the `default` case. give me a sec.

slowbackspace replied:
that spread in `custom` case is really a safety check, as the feature is intended only for internal use I wanted to prevent weird things happening to our beloved product team. I do not expect we are going to bother with migrations for this hidden top secret interal-only feature.

mroz22 commented on packages/suite/src/views/wallet/coinmarket/buy/offers/Offers/index.tsx:
```diff
@@ -43,12 +43,12 @@ const DividerMiddle = styled.div`
-    border: 1px solid ${colors.NEUE_STROKE_GREY};
-    background: ${colors.WHITE};
+    border: 1px solid ${props => props.theme.STROKE_GREY};
+    color: ${props => props.theme.BG_WHITE};
```
original 'background', new: 'color', color is doubled, is that expected?

slowbackspace replied:
good catch :) I'll probably ping invity team to take a look at their part of the app as they are more likely to notice unusual things

mroz22 commented on packages/suite/src/views/wallet/transactions/components/TransactionList/components/Pagination/index.tsx:
```diff
@@ -11,10 +11,10 @@ const Wrapper = styled.div`
+    color: ${props => (props.isActive ? props.theme.TYPE_WHITE : props.theme.TYPE_GREEN)};
     padding: 4px 8px;
-    border-radius: 2px;
+    border-radius: 2px;s
```
here is some left behind 's'.
this is actually something I don't like very much about styled components, you may miss `;` and checks still pass

slowbackspace replied:
good thing we have good reviewers :) I thought that would be caught by some of our linters, weird.

slowbackspace commented:
> Glanced over the functional changes of this PR and code looks mostly good to me. Just a few comments.
> 
> I found an issue with the toggling of the dark/light freezing the window for a second on Firefox (haven't checked in Chrome). This is from the performance tab when toggling it on and off.
> ![Screenshot from 2020-11-12 15-24-25](https://user-images.githubusercontent.com/30026749/98953217-b01d4980-24fc-11eb-9c70-48cb2f73bc44.png).
> 
I would expect some lag as react needs to rerender whole component tree (which is rather big), chrome is a bit faster with these things it seems (for me it is instant on Chrome, up to 1s lag in Firefox). Do you know some way how this can be optimised?

> Regarding the design, I'm not sure if we really need to have the dark-mode switch button in the nav bar. It feels to me like it's a setting that once enabled will rarely get fiddled with. Detection is automatic and I feel we should only have it in the settings.

You are correct, there won't be a switch in the nav bar, only in settings :) I was just waiting for product decision.

>I would also maybe suggest changing it into a list instead of a toggle and change the wording around theming. It would be nice to have more themes in the future (community contributions, secret unlockable or even accessibility focused theme) and would show also that we're open to the idea of adding more.

I agree, but right know the decision is to only include one dark mode, if in the future we will decide to support more themes it will be easy pretty easy to replace switch with select component, I find it more intuitive to have dark mode toggle than selector with 2 themes (light, dark). Let's get the basics of the dark mode right and later we will come up with more themes :)

> 
> I've also noticed analytics is missing. It would be interesting to see if people use themes or not.

Good points! @mroz22 will you take care of it please?
> 
> And lastly, some documentation would be really nice. That would be even more useful if we would consider the multi-theme way I described above. The community could use it to design their own themes.

Ohh, you probably mean documentation for the color palette in debug settings, that's is not supposed to be very user accessible. These settings are hidden from the user (It's really an ad-hoc solution). The purpose of it  is to provide a bit easier way to tinker our dark theme for product team (and devs I guess). Also I want to take time to refactor some of the color definitions, so it wouldn't make much sense to document it right now.

If you mean docs for the technical part then let me now the rough outline what would you like to see in it. I thought the implementation is pretty straight-forward, but ofc that may be just because I was the one doing it 😁 

> 
> Overall, nice work, I love it ❤️

❤️

mroz22 commented:
Regarding analytics, just please create an issue whenever you feel there is something worth reporting.

slowbackspace commented:
Thank you guys! I've fixed a few things and removed the toggle from navbar. I see invity managed to get some of their stuff merged so I'm gonna rebase later once we made sure that the fixups commits actually fixed things :)

heey, btw @keraf did you noticed fixed scrolling in firefox? 😁 also new sexy themed scrollbars (well, only their color changed, it's now based on selected theme)!

slowbackspace commented:
Ohh and I got an update from design, we are gonna use a bit lighter colors (🎉 🎉 🎉 ), so I'll update the definitions (one file change only, don't worry)

mroz22 commented:
Good please rebase and let's merge this hydra

keraf commented:
Let's not jump the gun on this one and wait at least for Monday before merging. There are still a few things to be looked at, especially the performance issues on Firefox. It would also be a good idea to let QA run the branch first (cc @trezor/qa).

slowbackspace commented:
@keraf  The perf issue you are talking about is only while you are actually switching themes, which most users don't do daily. If there is a room for optimisation then I am more than happy to do it, but I don't know how it could be improved. Our component tree is huge and it just might be the cause that rerendering  of everything is pretty heavy op, or do you see other problems?

There are tons of places that could use some optimizations I am not sure if action that user will be doing 2 times in their whole life is worth it to maybe save half a sec in one browser.

edit: Just to be clear, I am not saying that lag in firefox is not a bad for an UX, I just find weird we are making a blocker out of this. Rebasing this branch is not something I enjoy doing :)

slowbackspace commented:
Added some more commits, last 2 or 3 are adding new invity stuff that got added after the rebase

keraf commented:
Any action that freezes the whole render process for a second is bad UX, no matter how often the user does it. If it's fast in Chrome and not Firefox, there must be reason for that and we should be looking into it. We have some creeping performance problems and adding more on top of it is unwise in my opinion. I'll check out the branch and look into it.

slowbackspace commented:
@keraf attached gif of firefox when changing theme in settings, no perf issues here. As I already said, if you had slow experience in the dashboard, it might be because of our huge component tree that is taking some time to rerender (chrome is always way faster). If that't the case then fixing it is really out of scope of this PR (and I would still say it doesn't matter that much since the switch will be only in settings 😁). 
But ofc feel free to take a look at it, I just don't wanna keep rebasing this forever.

![Screen Recording 2020-11-13 at 15 01 03](https://user-images.githubusercontent.com/6961901/99080468-a578b800-25c1-11eb-85fd-b6e5d1dc183e.gif)

keraf commented:
I can constantly reproduce the slow down on the settings page on two different machines (both running Ubuntu 20.04) in Firefox. I just tried it in Chromium and it's totally broken there (the toggle is even slower to react and it never switches to dark mode). I haven't tested a prod build yet but we can check on the deployed environment on the CI passes.

The culprit according to the React profiler is the `GlobalStyle` component. Haven't dug too deep into it yet but will do so shortly.

slowbackspace commented:
@keraf  waait, you are testing the performance locally on dev builds? these are gonna be slow by design - https://reactjs.org/docs/optimizing-performance.html#use-the-production-build
also make sure you did git pull :) probably also clear storage if you tried this branch before (I did some incompatible changes since the first introduction of this branch)

the gif I posted is taken from https://suite.corp.sldev.cz/suite-web/feat/dark/web/

keraf commented:
Local dev builds are slower than prod builds, that's no contest, but not slow to this point. We should try to keep our development builds relatively fast or we'll end-up in situations where debugging is painful. I guess for this one it's not too bad apart from the startup dip. Checked in the deployed build, about 100ms freeze which is negligible imo. 

Edit: Checked local dev build with latest changes, freeze is less noticeable (< 1s). That's good enough for me.

tomexx commented:
First of all - very good job! 👍 🔥

+ I have couple of suggestions regarding the design:

1) I'd go for lighter color when hovering/selecting an item e.g. in dropdown select etc.
![Screenshot 2020-11-16 at 12 28 51](https://user-images.githubusercontent.com/2822756/99248323-8f106d80-2808-11eb-8506-ff27e66fd402.png)

2) I'd maybe adjust the background color of secondary buttons to a lighter (greyish) version. See now and my suggestion
![Screenshot 2020-11-16 at 12 30 11](https://user-images.githubusercontent.com/2822756/99248509-db5bad80-2808-11eb-9d72-3de868ce6b8e.png)
![Screenshot 2020-11-16 at 12 30 02](https://user-images.githubusercontent.com/2822756/99248529-e4e51580-2808-11eb-83cf-02ccf084a42c.png)

3) I'd lower the contranst on pure white elements, e.g. toggle switches. See my propsoal for the off state of a toggle switch
![Screenshot 2020-11-16 at 12 31 33](https://user-images.githubusercontent.com/2822756/99248997-aa2fad00-2809-11eb-81b0-6a4ed0195f11.png)

4) Transaction graph colors + graph tooltip colors looks a bit odd now also
![Screenshot 2020-11-16 at 12 32 58](https://user-images.githubusercontent.com/2822756/99248796-5624c880-2809-11eb-8475-0280e7a8b625.png)
![Screenshot 2020-11-16 at 12 33 06](https://user-images.githubusercontent.com/2822756/99248802-56bd5f00-2809-11eb-8d9a-235860969db9.png)

5) "Soon" tag seems to be too bright
![Screenshot 2020-11-16 at 12 35 13](https://user-images.githubusercontent.com/2822756/99248858-705ea680-2809-11eb-8fdc-3eb564bf40ea.png)

Please consider changing these (consult with designers) :)

slowbackspace commented:
> First of all - very good job! 👍 🔥
> 
> * I have couple of suggestions regarding the design:
> 
> 1. I'd go for lighter color when hovering/selecting an item e.g. in dropdown select etc.
> 2. I'd maybe adjust the background color of secondary buttons to a lighter (greyish) version. See now and my suggestion
> 4. Transaction graph colors + graph tooltip colors looks a bit odd now also
> 5. "Soon" tag seems to be too bright
> 
> Please consider changing these (consult with designers) :)

Thank you!
passed 1, 2 and 4 to our designer. I also have a few idea how to make things better so do not worry, this is not like we are gonna release it tomorrow and never improve. I would rather expect more changes to the design in upcoming weeks :)

3  - should be better now, it looks like our override for switch color never worked 🎉 
5 -  I did ad-hoc fix. I expect the soon badge will be gone before the release anyway.

I also fixed color for elements that haven't defined their own color and the color was also not defined anywhere in an ancestors (invity stuff)

slowbackspace commented:
I am gonna express my concerns once more, The sooner it will land into the develop the better as we'll all have more time to catch visually subtle things that could have real impact (eg. dark text on dark bg), although I think there won't be many of these. There is no way 1 person/QA can catch all in a few days. That's why I made this PR early in our release cycle.

The basics are done, it works and can be iteratively improved

✌️

prusnak commented:
I agree with Maros here. Let's get this in ASAP if it's not breaking anything and we can deal with the smaller visual changes later.
Ideally we should start another issue for collecting these ideas for improvements.
